File: python3/score_of_parentheses.py
"""
Space   : O(1)
Time    : O(n)
"""

class Solution:
    def scoreOfParentheses(self, S: str) -> int:
        ans = 0
        pwr = -1
        isAdded = False
        
        for l in S:
            if l == "(":
                pwr += 1 
                isAdded = False
            else:
                if not isAdded:
                    ans += 2 ** pwr
                    isAdded = True
                pwr -= 1
        
        return ans

File: golang/score_of_parentheses.go
package main

// Space	: O(1)
// Time		: O(n)

import (
	"math"
)

func scoreOfParentheses(S string) int {
	ans, pwr := 0, -1
	isAdded := false

	for _, letter := range S {
		if string(letter) == "(" {
			isAdded = false
			pwr++
		} else {
			if !isAdded {
				ans += int(math.Pow(2.0, float64(pwr)))
				isAdded = true
			}
			pwr--
		}
	}

	return ans
}
